sklearn/metrics/pairwise.py

```diff
@@ -585,6 +585,189 @@ def cosine_distances(X, Y=None):
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, n_samples)
+
+    Reference
+    -------
+    Gower, J.C., 1971, A General Coefficient of Similarity and Some of Its
+    Properties.
+
+    Notes
+    ------
+    The non-numeric features, and numeric feature ranges are determined from
+    X and not Y.
+
+    No support for sparse matrices.
+    """
+
+    if issparse(X) or issparse(Y):
+        raise TypeError("Sparse matrices are not supported for gower distance")
+
+    y_none = Y is None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X)
+
+    array_type = np.object
+
+    # this if is necessary as strangelly the validator is rejecting numeric
+    # arrays with NaN
+    if np.issubdtype(X.dtype, np.number) and \
+            (np.isfinite(X.sum()) or np.isfinite(X).all()):
+        array_type = type(np.zeros(1, X.dtype).flat[0])
+
+    X, Y = check_pairwise_arrays(X, Y, precomputed=False, dtype=array_type)
+
+    X, Y = check_pairwise_arrays(X, Y, dtype=np.object)
+
+    n_rows, n_cols = X.shape
+
+    if categorical_features is None:
+        categorical_features = np.zeros(n_cols, dtype=bool)
+        for col in range(n_cols):
+            # In numerical columns, None is converted to NaN,
+            # and the type of NaN is recognized as a number subtype
+            if not np.issubdtype(type(X[0, col]), np.number):
+                categorical_features[col] = True
+    else:
+        categorical_features = np.array(categorical_features)
+
+    if np.issubdtype(categorical_features.dtype, np.int):
+        new_categorical_features = np.zeros(n_cols, dtype=bool)
+        new_categorical_features[categorical_features] = True
+        categorical_features = new_categorical_features
+    else:
+        categorical_features = np.array(categorical_features, dtype=np.bool)
+
+    # Categorical columns
+    X_cat = X[:, categorical_features]
+
+    # Numerical columns
+    X_num = X[:, np.logical_not(categorical_features)]
+    ranges_of_numeric = None
+    max_of_numeric = None
+
+    # Calculates the normalized ranges and max values of numeric values
+    _, num_cols = X_num.shape
+    ranges_of_numeric = np.zeros(num_cols)
+    max_of_numeric = np.zeros(num_cols)
+    for col in range(num_cols):
+        col_array = X_num[:, col].astype(np.float32)
+        max = np.nanmax(col_array)
+        min = np.nanmin(col_array)
+
+        if np.isnan(max):
+            max = 0.0
+        if np.isnan(min):
+            min = 0.0
+        max_of_numeric[col] = max
+        ranges_of_numeric[col] = (1 - min / max) if (max != 0) else 0.0
+
+    # This is to normalize the numeric values between 0 and 1.
+    X_num = np.divide(X_num, max_of_numeric, out=np.zeros_like(X_num),
```

Comment:
Do you subtract off the min somewhere? Otherwise we're not scaling between 0 and 1.

You should probably be using a MinMaxScaler, but it doesn't yet support NaNs